mergify_engine/rules/__init__.py

```diff
@@ -75,55 +91,150 @@ async def __call__(self, obj: filter.GetAttrObjectT) -> bool:
         return self.match
 
     def get_attribute_name(self) -> str:
-        tree = self.partial_filter.tree.get("-", self.partial_filter.tree)
+        tree = typing.cast(filter.TreeT, self.partial_filter.tree)
+        tree = tree.get("-", tree)
         name = list(tree.values())[0][0]
         if name.startswith(filter.Filter.LENGTH_OPERATOR):
             return str(name[1:])
         return str(name)
 
 
 @dataclasses.dataclass
-class RuleConditions:
-    conditions: typing.List[RuleCondition]
+class RuleConditionGroup:
+    """This describe a group leafs of the `conditions:` tree linked by and or or."""
+
+    condition: dataclasses.InitVar[
+        typing.Dict[
+            typing.Literal["and", "or"],
+            typing.List[typing.Union["RuleConditionGroup", RuleCondition]],
+        ]
+    ]
+    limit_leaf_types: bool = True
+    operator: typing.Literal["and", "or"] = dataclasses.field(init=False)
+    conditions: typing.List[
+        typing.Union["RuleConditionGroup", RuleCondition]
+    ] = dataclasses.field(init=False)
     match: bool = dataclasses.field(init=False, default=False)
     _used: bool = dataclasses.field(init=False, default=False)
 
-    async def eval(self, obj: filter.GetAttrObjectT) -> bool:
+    def __post_init__(
+        self,
+        condition: typing.Dict[
+            typing.Literal["and", "or"],
+            typing.List[typing.Union["RuleConditionGroup", RuleCondition]],
+        ],
+    ) -> None:
+        if len(condition) != 1:
+            raise RuntimeError("Invalid condition")
+
+        self.operator, self.conditions = next(iter(condition.items()))
+        if not self.limit_leaf_types:
+            return
+        for cond in self.conditions:
+            if isinstance(cond, RuleCondition):
+                attribute = cond.get_attribute_name()
+                # TODO(sileht): remove this heuristics to allow this attribute everywhere
+                if attribute.partition("-")[0] in [
+                    "check",  # heuristic in merge/queue
+                    "status",  # heuristic in merge/queue
+                    "head",  # heuristic for hiding useless rules
+                    "base",  # heuristic for hiding useless rules
+                    "author",  # heuristic for hiding useless rules
+                    "merged_by",  # heuristic for hiding useless rules
+                    "merged",  # heuristic for detecting if pull request is merged by us
+                    "closed",  # heuristic for detecting if pull request is merged by us
```

Comment:
This is pretty bad; do you have a plan to remove this limitation in the end?

Reply:
I have found a solution for each one in further PR of the stack.